### PR TITLE
Rewrite Route and Middleware commands

### PR DESCRIPTION
**Pull request by TimCurwick**
Breaking change - Output of Get-WebRoute changed from dictionary to array of psobjects.

Other changes
1.	Polaris.psm1
a.	Removed unnecessary $m variable creation. (It isn’t used again.)
b.	.Module.OnRemove
i.	Added -ErrorAction to ignore existing bugs.
c.	New-WebRoute
i.	Updated help.
ii.	Simplified If conditional.
iii.	Added whitespace for improved readability.
iv.	Added validation set for parameter -Method.
v.	Mandatory = $True to parameters -ScriptBlock and -ScriptPath.
vi.	Added parameter sets ScriptBlock and ScriptPath.
vii.	Added handling to make leading solidus (/) optional.
viii.	Removed unnecessary and/or redundant conversions of $Scriptblock from string to scriptblock and back to string.
ix.	Replaced ThrowError with $PSCmdlet.WriteError. (ThrowError creates a dependency on the DSC module.)
x.	Added -Force switch. Added check for existing web route. (-Force causes overwrite, else non-terminating error.)
d.	Remove-WebRoute
i.	Updated help.
ii.	Simplified If conditional.
iii.	Added whitespace for improved readability.
iv.	Removed redundant positional parameter specifications (Parameters are positional by default if no parameter sets defined.)
v.	Removed Mandatory from parameters -Path and -Method.
vi.	Added handling for pipeline input, multiple values, and wildcard values.
e.	Get-WebRoute
i.	Updated help.
ii.	Simplified If conditional.
iii.	Added whitespace for improved readability.
iv.	Added conversion of dictionary to expanded psobjects, with Path, Method, and Scriptblock properties. – Breaking change
v.	Added optional parameters Path, Method. Each can take multiple values and PowerShell wildcards. Each defaults to ‘*’.
vi.	Added handling for pipeline input.
f.	New-StaticRoute
i.	Updated help.
ii.	Added whitespace for improved readability.
iii.	Removed redundant positional parameter specifications (Parameters are positional by default if no parameter sets defined.)
iv.	Moved $RoutePath.TrimEnd() out of loop to improve efficiency.
v.	Moved PowerShell version logic out of script template to simplify and improve efficiency of the route scriptblock.
vi.	Added parameter switch -Force. Value is passed through to New-WebRoute.
g.	New-RouteMiddleware
i.	Updated help.
ii.	Simplified If conditional.
iii.	Added whitespace for improved readability.
iv.	Added parameters sets ScriptBlock and ScriptPath.
v.	Removed unnecessary and/or redundant conversion of $Scriptblock from string to scriptblock and back to string.
vi.	Replaced ThrowError with $PSCmdlet.WriteError. (ThrowError creates a dependency on the DSC module.)
vii.	Added -Force switch. Added check for existing route middleware. (-Force causes overwrite, else non-terminating error.)
h.	Remove-RouteMiddleWare
i.	Updated help.
ii.	Simplified If conditional.
iii.	Added whitespace for improved readability.
iv.	Removed redundant positional parameter specification. (Parameters are positional by default if no parameter sets defined.)
v.	Removed Mandatory from parameter -Name.
vi.	Added handling for pipeline input, multiple values, and wildcard values.
i.	Get-RouteMiddleware
i.	Updated help.
ii.	Simplified If conditional.
iii.	Added whitespace for improved readability.
iv.	Removed redundant positional parameter specification.
v.	Set default parameter -Name value to ‘*’.
vi.	Added handling for parameter -Name wildcard values and multiple values.
vii.	Added handling for pipeline input.
j.	Start-Polaris
i.	Updated help.
ii.	Added whitespace for improved readability.
iii.	Removed redundant positional parameter specifications (Parameters are positional by default if no parameter sets defined.)
k.	Stop-Polaris
i.	Updated help.
ii.	Simplified If conditional.
iii.	Added whitespace for improved readability.
iv.	Removed redundant positional parameter specifications (Parameters are positional by default if no parameter sets defined.)
v.	Removed redundant check if $ServerContext is null. ($ServerContext can never be null.)
vi.	Removed try/catch which always fails.
l.	New-GetRoute
i.	Updated help.
ii.	Added whitespace for improved readability.
iii.	Added Mandatory = $True to parameters $ScriptBlock and $ScriptPath.
iv.	Added handling for new parameter sets on New-WebRoute.
v.	Added -Force switch. Value is passed to New-WebRoute.
m.	New-PostRoute
i.	Updated help.
ii.	Added whitespace for improved readability.
iii.	Added Mandatory = $True to parameters $ScriptBlock and $ScriptPath.
iv.	Added handling for new parameter sets on New-WebRoute.
v.	Added -Force switch. Value is passed to New-WebRoute.
n.	New-PutRoute
i.	Updated help.
ii.	Added whitespace for improved readability.
iii.	Added Mandatory = $True to parameters $ScriptBlock and $ScriptPath.
iv.	Added handling for new parameter sets on New-WebRoute.
v.	Added -Force switch. Value is passed to New-WebRoute.
o.	New-DeleteRoute
i.	Updated help.
ii.	Added whitespace for improved readability.
iii.	Added Mandatory = $True to parameters $ScriptBlock and $ScriptPath.
iv.	Added handling for new parameter sets on New-WebRoute.
v.	Removed redundant CreateNewPolarisIfNeeded. (It’s called within New-WebRoute.)
p.	Use-JsonBodyParserMiddleware
i.	Updated help.
ii.	Added -Force to New-RouteMiddleware
q.	CreateNewPolarisIfNeeded
i.	Simplified If conditional.
ii.	Simplified object creation.
iii.	Added whitespace for improved readability.
r.	Export-ModuleMember
i.	Removed redundant line continuations. (Line break okay after comma.)
2.	Polaris.psd1
a.	Set FunctionToExport to equal ‘*’. (Defined in .psm1.)
3.	/Tests
a.	Added Pester test files.
i.	New-WebRoute.Tests.ps1
ii.	Remove-WebRoute.Tests.ps1
iii.	Get-WebRoute.Tests.ps1
iv.	New-StatisRoute.Tests.ps1
v.	New-RouteMiddleware.Tests.ps1
vi.	Remove-RouteMiddleware.Tests.ps1
vii.	Get-RouteMiddleware.Tests.ps1
viii.	Helpers.Tests.ps1

### CODE REVIEW

**Comment by tylerl0706** [comment]
Fixed the merge conflict :) I'll continue looking at this. Thank you for such a great contribution!

File: test/Remove-RouteMiddleware.Tests.ps1 [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+
+        Test-RemoveMiddleware -Name 'Test3*'
+        }
+
+    It "Should accept Name from pipeline" {
+
+        $Name = 'Test4A', 'Test4B'
+
+        #  Get before state
+        $MiddlewareBefore   = Get-RouteMiddleware
```
Comment (tylerl0706) [inline test/Remove-RouteMiddleware.Tests.ps1]: Is there a reason you can't use the helper function (`Test-RemoveMiddleware`) you defined above for this test?

File: test/Remove-RouteMiddleware.Tests.ps1 [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+        $DeletedMiddlewareRemaining.Count | Should Be 0
+        $TotalRemainingMiddleware.Count   | Should Be $MiddlewareExpectedAfter
+        }
+
+    It "Should accept Name from pipeline variables" {
+
+        $Name = 'Test5*'
+
+        #  Get before state
+        $MiddlewareBefore   = Get-RouteMiddleware
```
Comment (tylerl0706) [inline test/Remove-RouteMiddleware.Tests.ps1]: Same comment about `Test-RemoveMiddleware`

File: test/Remove-WebRoute.Tests.ps1 [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+
+        #  Test after state
+        $DeletedRoutesRemaining.Count | Should Be 0
+        $TotalRemainingRoutes.Count   | Should Be $RoutesExpectedAfter
+        }
+
+    It "Should accept Path and Method from pipeline variables" {
+
+        #  Get before state
+        $RoutesBefore   = Get-WebRoute
```
Comment (tylerl0706) [inline test/Remove-WebRoute.Tests.ps1]: Is there a reason you can't use the helper function (`Test-RemoveRoute`) you defined above for this test?

File: test/Remove-WebRoute.Tests.ps1 [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+        }
+
+    It "Should accept multiple values for Method" {
+        Test-RemoveRoute -Path 'Test5' -Method 'GET', 'POST'
+        }
+
+    It "Should accept Path from pipeline" {
+
+        #  Get before state
+        $RoutesBefore   = Get-WebRoute
```
Comment (tylerl0706) [inline test/Remove-WebRoute.Tests.ps1]: Same comment about `Test-RemoveRoute`

File: Polaris.psm1 [hunk 5]
```diff
@@ -1,584 +1,807 @@
...
     [CmdletBinding()]
     param(
-        [Parameter(Mandatory=$True, Position=0)]
+        [Parameter( Mandatory = $True, Position = 0 )]
         [string]
         $Path,
 
-        [Parameter(Mandatory=$True, Position=1)]
+        [Parameter( Mandatory = $True, Position = 1 )]
+        [ValidateSet( 'GET', 'POST', 'PUT', 'DELETE' )]
```
Comment (tylerl0706) [inline Polaris.psm1]: nitpick: HEAD and PATCH are also valid HTTP Methods

**Comment by TimCurwick** [comment]
Tyler,

Module location - fixed
$Env:Temp - Changed to 'TestDrive:\' (I tried this originally, but I was
doing something wrong and it didn't work. Works now.)
Functions to export - fixed
Expected thrown errors not expected to be thrown - fixed
Helper functions in Remove tests - had to make them more complex, but fixed

Method validateset - Not yet changed. I would prefer we create a custom
enum to use as the parameter types. Easier to maintain. - OTH, should we be
able to handle non-standard methods? The downside is we lose the list to
select from in Intellisense, which was the reason I put in the validateset
in the first place.

Thanks,
Tim Curwick

**Comment by TimCurwick** [comment]
Tyler,

Looks like I have a failure in on the Pester tests. I'll take a look.

On the Linux side, my new module file reference doesn't work. i didn't just
put in ..\polaris.psd1 because that would fix it for your build test, but
would cause the test to fail if run from elsewhere. Why is this needed at
all? Why isn't the Polaris module in a valid module folder during build
testing?

Thanks,
Tim Curwick

**Comment by TimCurwick** [comment]
All tests are now passing with the exception of one that was a pre-existing bug unrelated to this pull request.

**Comment by tylerl0706** [comment]
Thanks for this @TimCurwick! I'm currently investigating the test failure. It wasn't failing before your PR but your PR isn't really doing anything that would cause the failure so it's strange.

**Comment by tylerl0706** [comment]
@TimCurwick merging this in :) I did a quick fix to get the tests working.
 
Thank you so much for this great PR!

**Pull request merged**